lib/iris/analysis/__init__.py

```diff
@@ -1472,6 +1473,46 @@ def _proportion(array, function, axis, **kwargs):
     return result
 
 
+def _lazy_max_run(array, axis=-1, **kwargs):
+    """
+    Lazily perform the calculation of maximum run lengths along the given axis
+    """
+    array = iris._lazy_data.as_lazy_data(array)
+    func = kwargs.pop("function", None)
+    if not callable(func):
+        emsg = "function must be a callable. Got {}."
+        raise TypeError(emsg.format(type(func)))
+    bool_array = da.ma.getdata(func(array))
```

Comment:
So thinking about the proper way to do this sent me down a bit of a rabbit hole which ended up in me making a bug report to numpy (https://github.com/numpy/numpy/issues/21389).

Ultimately, I don't have full confidence in the line `bool_array = da.ma.getdata(func(array))`. Using `getdata` fetches the values underneath the mask, which seem to work for the test cases, but I'm not sure it would work for all reasonable cases of `func`. I think `da.ma.filled(func(array), 0)` should be more robust here as it ensures that every masked point _must_ be False.

From doing a bit of a dive into numpy, it appears that inequalities, which are probably one of the most common types of `func` we can expect, still have some issues with masked arrays (e.g. https://github.com/numpy/numpy/issues/10092), so I think it's worth being as robust as possible here.

Reply:
Doesn't the line below cause us to ignore whatever applying `func` to the masking has done? I don't think we can have a problem with this unless we get an exception from applying `func` or what comes out doesn't have truthiness?

Reply:
Ah yes, I think you're right. I guess your aproach might actually be more robust in the edge case where `func` does not preserve the mask.